src/main/scala/tech/cryptonomic/conseil/Lorre.scala

```diff
@@ -450,6 +462,107 @@ object Lorre extends App with TezosErrors with LazyLogging with LorreAppConfig w
+      * @param votingData
+      * @return
+      */
+    def partitionByVote(votingData: Seq[VotingData]): (Seq[VotingData], Seq[VotingData], Seq[VotingData]) = {
+      val (yays, notYays) = votingData.partition(x => x.ballot.contains("yay"))
+      val (nays, notYaysOrNays) = notYays.partition(x => x.ballot.contains("nay"))
+      val (passes, _) = notYaysOrNays.partition(x => x.ballot.contains("pass"))
+      (yays, nays, passes)
+    }
+
+    /**
+      * * Helper function, takes a sequence of GovernanceData, calculates stake.
+      * @param votingData
+      * @return
+      */
+    def calculateStake(votingData: Seq[VotingData]): Future[BigDecimal] =
+      Future
+        .sequence(
+          votingData.map {
+            case VotingData(_, _, hash, _, _, _, address) =>
+              tezosNodeOperator.getAccountBalanceForBlock(BlockHash(hash), AccountId(address.get))
+          }
+        )
+        .map(balances => balances.reduceOption(_ + _).getOrElse(0))
+
+    /**
+      * * Helper function, takes a sequence of GovernanceData, processes aggregate fields.
+      * @param votingData
+      * @return
+      */
+    def processVoteAggregate(votingData: Seq[VotingData]): Future[TezosTypes.VoteAggregates] = {
+      val sampleVotingField = votingData.head
+      val timestamp = sampleVotingField.timestamp
+      val cycle = sampleVotingField.cycle
+      val level = sampleVotingField.level
+      val proposalHash = sampleVotingField.proposalHash
+      val (yays, nays, passes) = partitionByVote(votingData)
+      val yayCount = yays.length
+      val nayCount = nays.length
+      val passCount = passes.length
+      for {
+        yayStake <- calculateStake(yays)
+        nayStake <- calculateStake(nays)
+        passStake <- calculateStake(passes)
+        totalStake = yayStake + nayStake + passStake
+      } yield
+        VoteAggregates(
+          timestamp,
+          cycle,
+          level,
+          proposalHash,
+          yayCount,
+          nayCount,
+          passCount,
+          yayStake,
+          nayStake,
+          passStake,
+          totalStake
+        )
+    }
+
+    //get fields from db in Tezos Database Operations, use the block levels as input
+    db.run(TezosDb.fetchVotingBlockLevels).flatMap { levels =>
+      Source
+        .fromIterator(() => levels.toIterator)
+        .grouped(batchingConf.blockPageSize) // we use here block page size, because
+        .mapAsync(1) { blockLevels =>
+          val computeAndStore = for {
+            listOfVotingFields <- TezosDb.fetchVotingData(blockLevels.toSet)
+            ballotOperationsByClock = listOfVotingFields.groupBy(_.level).values
+            votes <- DBIO.from {
+              Future.traverse(ballotOperationsByClock.toSeq)(processVoteAggregate)
+            }
+            dbWrites <- TezosDb.writeVotes(votes.toList)
+          } yield dbWrites
+
+          db.run(computeAndStore).andThen {
```

Comment:
Might we wanna wait to fold all iterated levels in a single result here?

Failure cases might need to be handled consistently to get reasonable results even if one "batch" of data fails to save... what happens in such case?
The stream will probably fail... I guess.

Just thinking out my worries, not saying that it's not correct as it is now... I'm just looking for a clear idea of how we keep under control the failures.

Reply:
yes, but when it starts again it starts from the max votes level

Reply:
We just need to make sure that we don't have half-processed "in-flight" batches that will not be re-processed at next restart. 👍

Reply:
do you think adding `.transactionally` would do the job?

Reply:
I think there might not be anything more needed, my goal was to make sure that corner cases and failures were accounted for